pinot-core/src/main/java/com/linkedin/pinot/core/segment/creator/impl/inv/OffHeapBitmapInvertedIndexCreator.java

```diff
@@ -117,31 +118,36 @@ public OffHeapBitmapInvertedIndexCreator(File indexDir, int cardinality, int num
-    if (!spec.isSingleValueField()) {
-      origLengths = MmapUtils.allocateDirectByteBuffer(numDocs * INT_SIZE, null,
-          "lengths buffer to create bitmap index for " + columnName);
-      lengths = origLengths.asIntBuffer();
-    } else {
-      origLengths = null;
-    }
+    try {
+      // create buffers to store raw values
+      origValueBuffer = MmapUtils.allocateDirectByteBuffer(this.capacity * INT_SIZE, null,
+          "value buffer create bitmap index for " + columnName);
+      valueBuffer = origValueBuffer.asIntBuffer();
+      if (!spec.isSingleValueField()) {
+        origLengths = MmapUtils.allocateDirectByteBuffer(numDocs * INT_SIZE, null,
+            "lengths buffer to create bitmap index for " + columnName);
+        lengths = origLengths.asIntBuffer();
+      } else {
+        origLengths = null;
+      }
 
-    // create buffer to store posting lists
-    origPostingListBuffer = MmapUtils.allocateDirectByteBuffer(this.capacity * INT_SIZE, null,
-        "posting list buffer to bitmap index for " + columnName);
-    postingListBuffer = origPostingListBuffer.asIntBuffer();
-    origPostingListLengths = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
-        "posting list lengths buffer to bitmap index for " + columnName);
-    postingListLengths = origPostingListLengths.asIntBuffer();
-    origPostingListStartOffsets = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
-        "posting list start offsets buffer to bitmap index for " + columnName);
-    postingListStartOffsets = origPostingListStartOffsets.asIntBuffer();
-    origPostingListCurrentOffsets = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
-        "posting list current offsets buffer to bitmap index for " + columnName);
-    postingListCurrentOffsets = origPostingListCurrentOffsets.asIntBuffer();
+      // create buffer to store posting lists
+      origPostingListBuffer = MmapUtils.allocateDirectByteBuffer(this.capacity * INT_SIZE, null,
+          "posting list buffer to bitmap index for " + columnName);
+      postingListBuffer = origPostingListBuffer.asIntBuffer();
+      origPostingListLengths = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
+          "posting list lengths buffer to bitmap index for " + columnName);
+      postingListLengths = origPostingListLengths.asIntBuffer();
+      origPostingListStartOffsets = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
+          "posting list start offsets buffer to bitmap index for " + columnName);
+      postingListStartOffsets = origPostingListStartOffsets.asIntBuffer();
+      origPostingListCurrentOffsets = MmapUtils.allocateDirectByteBuffer(cardinality * INT_SIZE, null,
+          "posting list current offsets buffer to bitmap index for " + columnName);
+      postingListCurrentOffsets = origPostingListCurrentOffsets.asIntBuffer();
+    } catch (Exception e) {
+      releaseResources();
```

Comment:
Release resource only if encountered exception?

Reply:
Correct. If there is no exception, this object is created fine, and the resources are released in the seal() method.